src/SAWScript/SolverCache/LMDBOptDatabase.hs

```diff
@@ -0,0 +1,292 @@
+decodeMaybe :: (String -> Either String a) -> [String] ->
+               Either String (Maybe a)
+decodeMaybe f [l] = Just <$> f l
+decodeMaybe _ [] = Right Nothing
+decodeMaybe _ ls = Left $ "Expected at most one line, got: " ++ show (unlines ls)
+
+-- | Given a list of exactly one string, apply the given decoding function to it
+decodeSingle :: (String -> Either String a) -> [String] -> Either String a
+decodeSingle f [l] = f l
+decodeSingle _ ls = Left $ "Expected one line, got: " ++ show (unlines ls)
+
+-- | Given an empty list, return the given element
+decodeEmpty :: a -> [String] -> Either String a
+decodeEmpty b [] = Right b
+decodeEmpty _ ls = Left $ "Expected nothing, got: " ++ show (unlines ls)
+
+
+-- Helper functions for IO -----------------------------------------------------
+
+-- | Apply 'hGetLine' to the given 'Handle' until the given 'String' is
+-- received, then return all the prior collected results of 'hGetLine'
+hGetLinesUntil :: String -> Handle -> IO [String]
+hGetLinesUntil stop_str h = do
+  l <- hGetLine h
+  if l == stop_str then return []
+  else (l:) <$> hGetLinesUntil stop_str h
+
+-- | Apply 'hGetLine' with a 'timeout' of the given number of microseconds
+-- repeatedly until either 'hReady' returns 'False' or the operation times out
+hGetLinesTimeout :: Int -> Handle -> IO [String]
+hGetLinesTimeout t_micros h =
+  timeout t_micros (hGetLine h) >>= \case
+    Just l -> hReady h >>= \case
+      True -> (l:) <$> hGetLinesTimeout t_micros h
+      False -> return [l]
+    Nothing -> return []
+
+
+-- Internal definition of and operations on LMDBOptDatabase --------------------
+
+-- | A key-value database which is either stored in memory, or if the user has
+-- supplied a path (see 'setPath'/'getPath'), as an LMDB database. Keys are
+-- represented as 'ByteString's, but values can be of any JSON-serialisable type
+-- (see 'ToJSON'/'FromJSON'). This is implemented by connecting to a @python3@
+-- process, so Python must be installed to use this type. To use the LMDB
+-- backend, the Python LMDB bindings must also be installed.
+data LMDBOptDatabase a =
+  LMDBOptDatabase
+  { dbIOVar :: MVar (String, MVar [String])
+  , dbIOTID :: ThreadId
+  , dbErr   :: Handle
+  , dbShell :: ProcessHandle
+  }
+
+-- | The timeout to use when attempting to get lines from @stderr@
+dbInternalTimeout :: Int
+dbInternalTimeout = 100000
+
+-- | Open a new @LMDBOptDatabase@ by connecting to a @python3@ process
+dbOpen :: IO (LMDBOptDatabase a)
+dbOpen = withSystemTempFile "lmdb_opt_database.py" $ \fnm fh -> do
+  -- First, start a python3 process with a temp file containing the contents of
+  -- lmdb_opt_database.py and the "shell" option for lmdb_opt_database.py
+  -- TODO: Actually use pip to install this file so we don't have to use a temp
+  -- file every time we open a database
+  hPutStr fh lmdb_opt_database__py >> hClose fh
+  (mb_i, mb_o, mb_e, p) <- createProcess $
+    (shell $ "python3 " ++ fnm ++ " shell")
```

Comment:
Needing to have `python` installed is a new form of dependency, and one that may be surprising if you're not expecting it. I think it would be worth documenting this in the `:?` output for `enable_solver_cache`, as well as somewhere in the SAW manual.

Reply:
Good catch! Added better docs to `enable_solver_cache` and `set_solver_cache_path`, working on the SAW manual now.